be/src/olap/rowset/rowset_converter.cpp

```diff
@@ -0,0 +1,115 @@
+                                                  RowsetMetaPB* dst_rs_meta_pb) {
+    return _convert_rowset(src_rowset_meta, rowset_path, ALPHA_ROWSET, dst_rs_meta_pb);
+}
+
+OLAPStatus RowsetConverter::convert_alpha_to_beta(const RowsetMetaSharedPtr& src_rowset_meta,
+                                                  const std::string& rowset_path,
+                                                  RowsetMetaPB* dst_rs_meta_pb) {
+    return _convert_rowset(src_rowset_meta, rowset_path, BETA_ROWSET, dst_rs_meta_pb);
+}
+
+OLAPStatus RowsetConverter::_convert_rowset(const RowsetMetaSharedPtr& src_rowset_meta,
+                                            const std::string& rowset_path,
+                                            RowsetTypePB dst_type,
+                                            RowsetMetaPB* dst_rs_meta_pb) {
+    const TabletSchema& tablet_schema = _tablet_meta->tablet_schema();
+    RowsetWriterContext context;
+    context.rowset_id = src_rowset_meta->rowset_id();
+    context.tablet_uid = _tablet_meta->tablet_uid();
+    context.tablet_id = _tablet_meta->tablet_id();
+    context.partition_id = _tablet_meta->partition_id();
+    context.tablet_schema_hash = _tablet_meta->schema_hash();
+    context.rowset_type = dst_type;
+    context.rowset_path_prefix = rowset_path;
+    context.tablet_schema = &tablet_schema;
+    context.rowset_state = src_rowset_meta->rowset_state();
+    if (context.rowset_state == VISIBLE) {
+        context.version = src_rowset_meta->version();
+        context.version_hash = src_rowset_meta->version_hash();
+    } else {
+        context.txn_id = src_rowset_meta->txn_id();
+        context.load_id = src_rowset_meta->load_id();
+    }
+    std::unique_ptr<RowsetWriter> rowset_writer;
+    RETURN_NOT_OK(RowsetFactory::create_rowset_writer(context, &rowset_writer));
+    if (!src_rowset_meta->empty()) {
+        RowsetSharedPtr rowset;
+        RETURN_NOT_OK(RowsetFactory::create_rowset(&tablet_schema, rowset_path, src_rowset_meta, &rowset));
+        RowsetReaderSharedPtr rowset_reader;
+        RETURN_NOT_OK(rowset->create_reader(&rowset_reader));
+        std::vector<uint32_t> cids;
+        for (int i = 0; i < tablet_schema.num_columns(); ++i) {
+            cids.push_back(i);
+        }
+        DeleteHandler delete_handler;
+        RowsetReaderContext reader_context;
+        reader_context.reader_type = READER_ALTER_TABLE;
+        reader_context.tablet_schema = &tablet_schema;
+        reader_context.return_columns = &cids;
+        reader_context.seek_columns = &cids;
+        reader_context.delete_handler = &delete_handler;
+        RETURN_NOT_OK(rowset_reader->init(&reader_context));
+        // convert
+        RowBlock* row_block = nullptr;
+        RowCursor row_cursor;
+        row_cursor.init(tablet_schema);
+        while (true) {
+            if (row_block == nullptr || !row_block->has_remaining()) {
```

Comment:
I think read logic should not exposed outside of reader.cpp

Reply:
RowsetConverter is in olap/rowset. It is used for single rowset, no related with data model, so I think it is ok. And more I think rowset interface can be used directly, like schema change.